### PR TITLE
Implement P3 elements

### PR DESCRIPTION
**Pull request by drwells**
I'd like to try out P3 elements in a project. Implementing `FE_SimplexDGP<dim>(3)` was easy, but the continuous implementation revealed a bunch of missing features in our tet and triangle orientation code. The 2D version should now work but I need @peterrum's help filling out some more tetrahedral orientation tables. In particular, right now if we print out support point locations we get inconsistent results between elements in 3D:

```
[ ... ]

DEAL::DoF 8
DEAL::  support point 0.00000 0.666667 0.00000
DEAL::DoF 9
DEAL::  support point 0.00000 0.333333 0.00000

[...]

DEAL::DoF 8
DEAL::  support point -5.55112e-17 0.333333 0.00000
DEAL::DoF 9
DEAL::  support point -1.11022e-16 0.666667 0.00000
```

### CODE REVIEW

**Comment by peterrum** [comment]
I thinks something like this should work: https://github.com/peterrum/dealii/blob/2b9e707f9a4aea2278b01b34dff8d973bfa61f00/include/deal.II/grid/reference_cell.h#L2350-L2378

I used the following script to generate the table. But I am not sure if this is correct (your test does not seem to be happy). @drwells maybe you could use this as a starting point!?
```python
temp = []

for face in [[0,1,2], [1,0,3],[0,2,3],[2,1,3]]:
	temp_orientations = []
	for face_orientation in [[0,2,1], [0,1,2],[2,1,0],[1,2,0],[1,0,2],[2,0,1]]:
		temp_lines = []
		for line in [[0,1],[1,2],[0,2]]:
			
			ids = [[face[o] for o in face_orientation][l] for l in line]
			if ids[0] < ids[1]:
				temp_lines = temp_lines + ["true"]
			else:
				temp_lines = temp_lines + ["false"]

		temp_orientations = temp_orientations + [temp_lines]
	
	temp = temp + [temp_orientations]

print "{{" + ", ".join(["{{" + ", ".join(["{{" + ", ".join(a) + "}}" for a in face]) + "}}" for face in temp]) + "}}"
```

**Comment by drwells** [comment]
I ended up using

```python
temp = []

# vertices in each face
for face in [[0,1,2], [1,0,3], [0,2,3], [2,1,3]]:
    print("face", face)
    temp_orientations = []
    for face_orientation in [[0,2,1], [0,1,2], [1, 0, 2], [1, 2, 0], [2, 1, 0], [2, 0, 1]]:
        temp_lines = []
        for line in [[0,1], [1,2], [0,2]]:

            ids = [[face[o] for o in face_orientation][l] for l in line]
            if (face_orientation == [0, 1, 2]):
                print(ids)
            if ids[0] < ids[1]:
                temp_lines = temp_lines + ["true"]
            else:
                temp_lines = temp_lines + ["false"]

        temp_orientations = temp_orientations + [temp_lines]

    temp = temp + [temp_orientations]

print("{{" + ", ".join(["{{" + ", ".join(["{{" + ", ".join(a) + "}}" for a in face]) + "}}" for face in temp]) + "}}")
```

which works but handles rotation inconsistently incorrectly - I'm still missing something for the more-correct `RO` counter-clockwise and `R` clockwise convention.

**Comment by peterrum** [comment]
I am wondering why you have swapped entries 2 and 4!?

File: include/deal.II/grid/connectivity.h [hunk 1]
```diff
@@ -195,7 +195,7 @@ namespace internal
         if (d == 1)
           {
             static const dealii::ndarray<unsigned int, 3, 2> table = {
-              {{{0, 1}}, {{1, 2}}, {{2, 0}}}};
```
Comment (peterrum) [inline include/deal.II/grid/connectivity.h:198]: I guess we should extract these changes in a separate PR.

File: include/deal.II/grid/reference_cell.h [hunk 2]
```diff
@@ -1455,6 +1468,62 @@ ReferenceCell::standard_line_to_face_and_line_index(
...
+  AssertIndexRange(line, n_lines());
+
+  if (*this == ReferenceCells::Vertex)
+    return vertex;
+  else if (*this == ReferenceCells::Line)
+    return vertex;
+  else if (*this == ReferenceCells::Quadrilateral)
+    return GeometryInfo<2>::line_to_cell_vertices(line, vertex);
+  else if (*this == ReferenceCells::Hexahedron)
+    return GeometryInfo<3>::line_to_cell_vertices(line, vertex);
```
Comment (bangerth) [inline include/deal.II/grid/reference_cell.h:1485]: We want to get rid of `GeometryInfo` in favor of `ReferenceCell`. Would you be willing to copy the implementation of these functions here?

**Comment by bangerth** [comment]
This patch has acquired merge conflicts. Would you mind updating it?

Separately, if you don't think it's ready yet, feel free to break out the parts that could stand on their own (such as the changes to `ReferenceCell`).

**Comment by drwells** [comment]
I still need to figure out the issue with tetrahedral line orientations before the P3 element can be used in 3D. Separating out the rest is a good idea - I'll do that!

File: tests/simplex/simplex_project.output [hunk 3]
```diff
@@ -0,0 +1,4521 @@
...
+DEAL::face 0: 1, 3, 2: orientation = 1
+DEAL::  line 0: 1, 3: orientation = 1
+DEAL::  line 1: 3, 2: orientation = 1
+DEAL::  line 2: 2, 1: orientation = 0
+DEAL::face 1: 3, 1, 0: orientation = 1
+DEAL::  line 0: 1, 3: orientation = 1
+DEAL::  line 1: 1, 0: orientation = 1
+DEAL::  line 2: 3, 0: orientation = 0
+DEAL::face 2: 1, 2, 0: orientation = 1
+DEAL::  line 0: 2, 1: orientation = 1
```
Comment (drwells) [inline tests/simplex/simplex_project.output:1852]: I think I've narrowed it down to this case (two tetrahedra in weird orientations): according to the first face, the `(2, 1)` line is flipped, but according to the third face, it is not flipped. Hence - based on which face we look up information on we get different orientations, which mess up line DoFs.

**Comment by drwells** [comment]
OK - it's possible to consistently orient faces in the way we need so that everything works, but we need to use a completely different tetrahedral numbering scheme. Some resulting TODOs:
- [ ] normal and tangential vectors
- [ ] isotropic refinement
- [ ] update triangles to use a similar convention
- [ ] remove extra face argument (and move to header)
- [ ] documentation

**Comment by drwells** [comment]
Peter and I are rearranging the contents of this PR into other places.
